File: Decompanion/decompanion/dialogs/action_selection_dialog.py
# action_selection_dialog.py

import threading
import javax.swing as swing
from javax.swing import JFrame, JPanel, JCheckBox, JButton, BoxLayout, JLabel
from java.awt import BorderLayout

# Dialog Class
class ActionSelectionDialog(JFrame):
    """
    Dialog for selecting the actions to perform.
    """
    def __init__(self, on_selection_complete):
        super(ActionSelectionDialog, self).__init__("Select Actions to Perform")
        self.selected_actions = []
        self.on_selection_complete = on_selection_complete
        self.init_ui()

    def init_ui(self):
        try:
            panel = JPanel()
            panel.setLayout(BoxLayout(panel, BoxLayout.Y_AXIS))

            self.add_instruction_label(panel)
            self.add_checkboxes(panel)
            self.add_buttons(panel)

            self.getContentPane().add(panel, BorderLayout.CENTER)
            self.setSize(300, 200)
            self.setLocationRelativeTo(None)
            self.setDefaultCloseOperation(JFrame.DISPOSE_ON_CLOSE)
            self.setVisible(True)
        except Exception as e:
            print "Error initializing ActionSelectionDialog UI: {}".format(e)

    def add_instruction_label(self, panel):
        try:
            instruction_label = JLabel("Select the actions you want to perform:")
            instruction_label.setToolTipText("Choose the desired actions")
            panel.add(instruction_label)
        except Exception as e:
            print "Error adding instruction label: {}".format(e)

    def add_checkboxes(self, panel):
        try:
            self.rename_retype_checkbox = JCheckBox("Rename and Retype Variables")
            self.rename_retype_checkbox.setSelected(True)
            self.rename_retype_checkbox.setToolTipText("Rename variables and update their types")
            panel.add(self.rename_retype_checkbox)

            self.explanation_checkbox = JCheckBox("Get Function Explanation")
            self.explanation_checkbox.setSelected(False)
            self.explanation_checkbox.setToolTipText("Obtain explanations for functions")
            panel.add(self.explanation_checkbox)

            self.line_comments_checkbox = JCheckBox("Add Line Comments")
            self.line_comments_checkbox.setSelected(False)
            self.line_comments_checkbox.setToolTipText("Add comments to lines of code")
            panel.add(self.line_comments_checkbox)
        except Exception as e:
            print "Error adding checkboxes: {}".format(e)

    def add_buttons(self, panel):
        try:
            button_panel = JPanel()
            ok_button = JButton("OK")
            ok_button.addActionListener(lambda e: self.ok())
            ok_button.setToolTipText("Confirm action selection")
            cancel_button = JButton("Cancel")
            cancel_button.addActionListener(lambda e: self.cancel())
            cancel_button.setToolTipText("Cancel action selection")
            button_panel.add(ok_button)
            button_panel.add(cancel_button)
            self.getContentPane().add(button_panel, BorderLayout.SOUTH)
        except Exception as e:
            print "Error adding buttons to ActionSelectionDialog: {}".format(e)

    def ok(self):
        try:
            self.selected_actions = []
            if self.rename_retype_checkbox.isSelected():
                self.selected_actions.append('rename_retype')
            if self.explanation_checkbox.isSelected():
                self.selected_actions.append('explanation')
            if self.line_comments_checkbox.isSelected():
                self.selected_actions.append('line_comments')
            self.on_selection_complete(self.selected_actions)
            self.dispose()
        except Exception as e:
            print "Error in OK action of ActionSelectionDialog: {}".format(e)

    def cancel(self):
        try:
            self.selected_actions = []
            self.on_selection_complete(self.selected_actions)
            self.dispose()
        except Exception as e:
            print "Error in Cancel action of ActionSelectionDialog: {}".format(e)

# Helper Functions
def show_action_select_dialog():
    """
    Displays the ActionSelectionDialog and waits for user interaction.

    Returns:
        list: The list of selected actions or an empty list if none selected.
    """
    selected_actions = []
    dialog_complete = threading.Event()

    def on_selection(selected):
        selected_actions.extend(selected)
        dialog_complete.set()

    def create_dialog():
        ActionSelectionDialog(on_selection)

    swing.SwingUtilities.invokeLater(create_dialog)
    dialog_complete.wait()
    return selected_actions

File: Decyx/decyx/dialogs/ActionSelectionDialog.java
package decyx.dialogs;

import javax.swing.*;
import java.awt.*;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CountDownLatch;

/**
 * Dialog for selecting the actions to perform (rename/retype, explanation, line comments).
 */
public class ActionSelectionDialog extends JFrame {

    private final List<String> selectedActions = new ArrayList<>();
    private final Runnable onComplete;
    private JCheckBox renameRetypeCheckbox;
    private JCheckBox explanationCheckbox;
    private JCheckBox lineCommentsCheckbox;

    public ActionSelectionDialog(Runnable onComplete) {
        super("Select Actions to Perform");
        this.onComplete = onComplete;
        initUI();
    }

    private void initUI() {
        try {
            JPanel panel = new JPanel();
            panel.setLayout(new BoxLayout(panel, BoxLayout.Y_AXIS));

            addInstructionLabel(panel);
            addCheckboxes(panel);
            addButtons(panel);

            getContentPane().add(panel, BorderLayout.CENTER);
            setSize(300, 200);
            setLocationRelativeTo(null);
            setDefaultCloseOperation(JFrame.DISPOSE_ON_CLOSE);
            setVisible(true);
        } catch (Exception e) {
            System.out.println("Error initializing ActionSelectionDialog UI: " + e.getMessage());
        }
    }

    private void addInstructionLabel(JPanel panel) {
        JLabel instructionLabel = new JLabel("Select the actions you want to perform:");
        instructionLabel.setToolTipText("Choose the desired actions");
        panel.add(instructionLabel);
    }

    private void addCheckboxes(JPanel panel) {
        renameRetypeCheckbox = new JCheckBox("Rename and Retype Variables");
        renameRetypeCheckbox.setSelected(true);
        renameRetypeCheckbox.setToolTipText("Rename variables and update their types");
        panel.add(renameRetypeCheckbox);

        explanationCheckbox = new JCheckBox("Get Function Explanation");
        explanationCheckbox.setSelected(false);
        explanationCheckbox.setToolTipText("Obtain explanations for functions");
        panel.add(explanationCheckbox);

        lineCommentsCheckbox = new JCheckBox("Add Line Comments");
        lineCommentsCheckbox.setSelected(false);
        lineCommentsCheckbox.setToolTipText("Add comments to lines of code");
        panel.add(lineCommentsCheckbox);
    }

    private void addButtons(JPanel panel) {
        JPanel buttonPanel = new JPanel();
        JButton okButton = new JButton("OK");
        okButton.addActionListener(e -> ok());
        okButton.setToolTipText("Confirm action selection");
        JButton cancelButton = new JButton("Cancel");
        cancelButton.addActionListener(e -> cancel());
        cancelButton.setToolTipText("Cancel action selection");
        buttonPanel.add(okButton);
        buttonPanel.add(cancelButton);
        getContentPane().add(buttonPanel, BorderLayout.SOUTH);
    }

    private void ok() {
        selectedActions.clear();
        if (renameRetypeCheckbox.isSelected()) {
            selectedActions.add("rename_retype");
        }
        if (explanationCheckbox.isSelected()) {
            selectedActions.add("explanation");
        }
        if (lineCommentsCheckbox.isSelected()) {
            selectedActions.add("line_comments");
        }
        onComplete.run();
        dispose();
    }

    private void cancel() {
        selectedActions.clear();
        onComplete.run();
        dispose();
    }

    public List<String> getSelectedActions() {
        return selectedActions;
    }

    /**
     * Displays the dialog and waits for user interaction.
     *
     * @return the list of selected action strings
     */
    public static List<String> showDialog() {
        CountDownLatch latch = new CountDownLatch(1);
        ActionSelectionDialog[] dialogHolder = new ActionSelectionDialog[1];

        SwingUtilities.invokeLater(() -> {
            dialogHolder[0] = new ActionSelectionDialog(latch::countDown);
        });

        try {
            latch.await();
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }

        return dialogHolder[0] != null ? dialogHolder[0].getSelectedActions() : new ArrayList<>();
    }
}
